brainglobe_utils/citation/format.py

```diff
@@ -0,0 +1,143 @@
+    Required fields are checked for on instantiation.
+    Missing optional fields will be set to None if not provided.
+
+    Parameters
+    ----------
+    information : Dict[str, str]
+        yaml-processed content of CITATION.cff.
+    warn_on_not_used: bool, default = False
+        If True, warn the user about the fields in the information
+        input that are not required nor optional, so are ignored.
+
+    Attributes
+    ----------
+    author_separator : str, default = ", "
+        Separator to use between authors in the event the reference
+        has more than one author provided.
+    indent_character : str, default = ' ' * 4
+        Character sequence to use for indentation.
+    required : ClassVar[List[str]]
+        Required fields for the entry to be created.
+    optional : ClassVar[List[str]]
+        Optional fields that the entry may possess.
+    """
+
+    indent_character: str = " " * 4
+    author_separator: ClassVar[str] = ", "
+    required: ClassVar[List[str]]
+    optional: ClassVar[List[str]]
+
+    # mypy type-hints
+    authors: Union[str, Dict[str, str], List[Dict[str, str]]]
+
+    @classmethod
+    def entry_type(cls) -> str:
+        """
+        The reference format type; article, book, software, etc.
+        This will be inherited by child classes and used to infer
+        which format the citation.cff files are providing.
+        """
+        return cls.__name__.lower()
+
+    def __init__(
+        self,
+        information: Dict[str, Any],
+        warn_on_not_used: bool = False,
+    ) -> None:
+        """ """
+        # So we don't delete information that we may need in other places
+        # (C++ memory sharing rights plz Python)
+        information = information.copy()
+
+        # Add all the information we need
+        for key, value in information.items():
+            if key in self.required or key in self.optional:
+                setattr(self, key.replace("-", "_"), value)
+            elif warn_on_not_used:
+                warnings.warn(
+                    f"The key {key} is not used for entries of type "
+                    f"{self.entry_type()}",
+                    UserWarning,
+                )
+
+        # Check that all required information is populated
+        for required_field in self.required:
+            if not hasattr(self, required_field):
+                raise KeyError(
+                    f"Did not receive value for required key: {required_field}"
+                )
+        # Optional fields should be set to None so that checks against
+        # them produce nothing and evaluated to False
+        for optional_field in self.optional:
+            if not hasattr(self, optional_field.replace("-", "_")):
+                setattr(self, optional_field.replace("-", "_"), None)
+
+        # If we have an authors field, we will need to parse the
+        # dictionary input into the string that BibTex is expecting
```

Comment:
Python class attributes cannot have "-" in their names, so this is more general than just BibTex (although you're correct in that this also means we don't have to worry about doing this _again_ when we write the keys in BibTex entries).

`setattr(self, string)` is effectively setting the property `self.{string}`. So if `string = "author-name"` for example, `setattr(self, string)` is trying to set `self.author-name` which will error since the `-` will be interpreted as an operation.

Reply:
Yeah we can probably do without the comment as you say. 

Originally I wrote the `BibTexEntry` class first, then later refactored the `Format` class out of it to recycle the instantiation/author-prep code for `TextCitation` (and any other formats we might have). Will do a quick scan for any more general cases that still refer to their bibtex counterpart